src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -178,20 +301,499 @@ func (s *commitLogSource) ReadData(
+
+			if mostRecentCompleteSnapshotForShardBlock.snapshotTime.Equal(blockStart) ||
+				// Should never happen
+				mostRecentCompleteSnapshotForShardBlock.isZero() {
+				// There is no snapshot file for this time, and even if there was, there would
+				// be no point in reading it. In this specific case its not an error scenario
+				// because the fact that snapshotTime == blockStart means we already accounted
+				// for the fact that this snapshot did not exist when we were deciding which
+				// commit logs to read.
+				s.log.Infof(
+					"No snapshots for shard: %d and blockStart: %d",
+					shard, blockStart.Unix())
+				continue
+			}
+
+			shardResult, err = s.bootstrapShardBlockSnapshot(
+				nsID, shard, blockStart, metadataOnly, shardResult, allSeriesSoFar, blockSize,
+				snapshotFiles, mostRecentCompleteSnapshotForShardBlock)
+			if err != nil {
+				return shardResult, err
+			}
+		}
+	}
+
+	if shardResult == nil {
+		shardResult = result.NewShardResult(0, s.opts.ResultOptions())
+	}
+	return shardResult, nil
+}
+
+func (s *commitLogSource) bootstrapShardBlockSnapshot(
+	nsID ident.ID,
+	shard uint32,
+	blockStart time.Time,
+	metadataOnly bool,
+	shardResult result.ShardResult,
+	allSeriesSoFar *result.Map,
+	blockSize time.Duration,
+	snapshotFiles fs.FileSetFilesSlice,
+	mostRecentCompleteSnapshot snapshotMetadata,
+) (result.ShardResult, error) {
+	var (
+		bOpts      = s.opts.ResultOptions()
+		blOpts     = bOpts.DatabaseBlockOptions()
+		blocksPool = blOpts.DatabaseBlockPool()
+		bytesPool  = blOpts.BytesPool()
+		fsOpts     = s.opts.CommitLogOptions().FilesystemOptions()
+		idPool     = s.opts.CommitLogOptions().IdentifierPool()
+	)
+
+	// Bootstrap the snapshot file
+	reader, err := s.newReaderFn(bytesPool, fsOpts)
+	if err != nil {
+		return shardResult, err
+	}
+
+	err = reader.Open(fs.DataReaderOpenOptions{
+		Identifier: fs.FileSetFileIdentifier{
+			Namespace:   nsID,
+			BlockStart:  blockStart,
+			Shard:       shard,
+			VolumeIndex: mostRecentCompleteSnapshot.ID.VolumeIndex,
+		},
+		FileSetType: persist.FileSetSnapshotType,
+	})
+	if err != nil {
+		return shardResult, err
+	}
+
+	s.log.Infof(
+		"Reading snapshot for shard: %d and blockStart: %d and volume: %d",
+		shard, blockStart.Unix(), mostRecentCompleteSnapshot.ID.VolumeIndex)
+	for {
+		var (
+			id               ident.ID
+			tagsIter         ident.TagIterator
+			data             checked.Bytes
+			expectedChecksum uint32
+		)
+		if metadataOnly {
+			id, tagsIter, _, _, err = reader.ReadMetadata()
+		} else {
+			id, tagsIter, data, expectedChecksum, err = reader.Read()
+		}
+		if err != nil && err != io.EOF {
+			return shardResult, err
+		}
+
+		if err == io.EOF {
+			break
+		}
+
+		var tags ident.Tags
+		if tagsIter.Remaining() > 0 {
+			tags, err = convert.TagsFromTagsIter(id, tagsIter, idPool)
+			if err != nil {
+				return shardResult, fmt.Errorf("unable to decode tags: %v", err)
+			}
+		}
+		tagsIter.Close()
```

Comment:
Maybe move this below the `existing, ...` check so we don't always decode the tags?

Reply:
Thats a good call, although before I do that can you remind me how are guarantees work if someone indexes the same series with different tags across different index blocks? What if they try and index it twice within the same index block with different tags also?

Reply:
done